sdk/azcore/tracing/tracing.go

```diff
@@ -189,3 +237,159 @@ type Attribute struct {
+type SpanContext struct {
+	spanID     SpanID
+	traceID    TraceID
+	traceFlags TraceFlags
+	traceState TraceState
+	remote     bool
+}
+
+// SpanID returns the SpanID from the SpanContext.
+func (sc SpanContext) SpanID() SpanID {
+	return sc.spanID
+}
+
+// TraceID returns the TraceID from the SpanContext.
+func (sc SpanContext) TraceID() TraceID {
+	return sc.traceID
+}
+
+// TraceFlags returns the flags from the SpanContext.
+func (sc SpanContext) TraceFlags() TraceFlags {
+	return sc.traceFlags
+}
+
+// TraceState returns the TraceState from the SpanContext.
+func (sc SpanContext) TraceState() TraceState {
+	return sc.traceState
+}
+
+// IsRemote indicates whether the SpanContext represents a remotely-created Span.
+func (sc SpanContext) IsRemote() bool {
+	return sc.remote
+}
+
+// SpanContextConfig contains mutable fields usable for constructing an immutable SpanContext.
+type SpanContextConfig struct {
+	TraceID    TraceID
+	SpanID     SpanID
+	TraceFlags TraceFlags
+	TraceState TraceState
+	Remote     bool
+}
+
+// NewSpanContext constructs a SpanContext using values from the provided SpanContextConfig.
+func NewSpanContext(config SpanContextConfig) SpanContext {
+	return SpanContext{
+		traceID:    config.TraceID,
+		spanID:     config.SpanID,
+		traceFlags: config.TraceFlags,
+		traceState: config.TraceState,
+		remote:     config.Remote,
+	}
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// PropagatorImpl contains the implementation for the Propagator.
+type PropagatorImpl struct {
+	// Inject contains the implementation for the Propagator.Inject method.
+	Inject func(ctx context.Context, carrier Carrier)
+	// Extract contains the implementation for the Propagator.Extract method.
+	Extract func(ctx context.Context, carrier Carrier) context.Context
+	// Fields contains the implementation for the Propagator.Fields method.
+	Fields func() []string
+}
+
+// Propagator is used to extract and inject context data from and into messages exchanged by applications.
+type Propagator struct {
+	impl PropagatorImpl
+}
+
+// Inject injects the span context into the carrier.
+func (p Propagator) Inject(ctx context.Context, carrier Carrier) {
+	if p.impl.Inject != nil {
+		p.impl.Inject(ctx, carrier)
+	}
+}
+
+// Extract extracts the span context from the carrier.
+func (p Propagator) Extract(ctx context.Context, carrier Carrier) context.Context {
+	if p.impl.Extract != nil {
+		return p.impl.Extract(ctx, carrier)
+	}
+	return ctx
+}
+
+// Fields returns the fields that the propagator can inject and extract.
+func (p Propagator) Fields() []string {
+	if p.impl.Fields != nil {
+		return p.impl.Fields()
+	}
+	return nil
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Carrier is the storage medium used by the Propagator.
+type Carrier interface {
```

Comment:
Interfaces are brittle and don't version well (i.e. adding a method to an interface is a breaking change), so this makes me nervous. It's also why we use this struct/func pattern in this package as we can add new `func` fields in a non-breaking way.

Reply:
That's a good point, I will update to make this a struct